### PR TITLE
internal: change generic parameter order

### PR DESCRIPTION
**Pull request by lowr**
tl;dr: This PR changes the `Substitution` for trait items and methods like so:

```rust
trait Trait<TP, const CP: usize> { // note the implicit Self as first parameter
  type Type<TC, const CC: usize>;
  fn f<TC, const CC: usize>() {}
}
impl<TP, const CP: usize> S {
  fn f<TC, const CC: usize>() {}
}
```

- before this PR: `[Self, TP, CP, TC, CC]` for each trait item, `[TP, CP, TC, CC]` for `S::f`
- after this PR: `[TC, CC, Self, TP, CP]` for each trait item, `[TC, CC, TP, CP]` for `S::f`

---

This PR swaps the generic parameters/arguments of an item and its parent. This is to fulfill [chalk's expectation](https://github.com/rust-lang/chalk/blob/d875af0ff196dd6430b5f5fd87a640fa5ab59d1e/chalk-solve/src/rust_ir.rs#L498-L502) on the order of generic arguments in `Substitution`s for generic associated types and it's one step forward for GATs support (hopefully). Although chalk doesn't put any constraint for other items, it feels more natural to get everything aligned than special casing GATs.

One complication is that `TyBuilder` now demands its users to pass in parent's `Substitution` upon construction unless it's obvious that the the item has no parent (e.g. an ADT never has parent). All users *should* already know the parent of the item in question, and without this, it cannot be easily reasoned about whether we're pushing the argument for the item or for its parent.

Some additional notes:
- f8f5a5ea5788a846013545d63c9b46fd70cc4f7c: This isn't related to the change, but I felt it's nicer.

- cda037c5c688b902d4d5cadf0032859b403fc6c8: There's one major change here other than the generic param order: Default arguments are now bound by the same `Binder` as the item in question rather than a `Binder` limited to parameters they can refer to (i.e. arguments that syntactically appear before them). Now that the order of generic parameters is changed, it would be somewhat complicated to make such `Binder`s as before, and the "full" `Binder`s shouldn't be a problem because we already make sure that the default arguments don't refer to the generic arguments after them with `fallback_bound_vars()`.

- 79094e8a0730d3d3d99be775645544d7449bd4b8: This is split from aafea391486a52c899e515e65dc97cdb6540756b to make it easy to revert if it turns out that the GATs with const generics panic is actually not resolved with this PR. cc #11878 #11957

### CODE REVIEW

**Comment by Veykril** [comment]
This immediately broke metrics 😅 https://github.com/rust-lang/rust-analyzer/actions/runs/3174097813/jobs/5170473829

**Comment by lowr** [comment]
I should really remember to run analysis-stats before submitting big changes :confused: 

It seems I messed up generics for enum variants. It shouldn't take too long to fix but I'm fine with reverting this PR if it's too problematic.

**Comment by Veykril** [comment]
if you have the time and want to try tackling it today feel free to, otherwise I'd say lets revert it as I believe this to break a lot of things for people

**Comment by lowr** [comment]
Yeah agreed. I'm going to make attempt but will let you know (either here or on zulip) if I can't seem to fix in the time frame.
